python_modules/dagster/dagster/_core/definitions/time_window_partition_mapping.py

```diff
@@ -273,33 +295,117 @@ def _map_partitions(
+                )
+            )
+        ):
+            return UpstreamPartitionsResult(
+                from_partitions_subset.with_partitions_def(to_partitions_def), []
+            )
+
+        # Daily to hourly
+        from_last_partition_window = from_partitions_def.get_last_partition_window(current_time)
+        to_last_partition_window = to_partitions_def.get_last_partition_window(current_time)
+        if (
+            from_partitions_def.cron_schedule == "0 0 * * *"
+            and to_partitions_def.cron_schedule == "0 * * * *"
+            and (
+                from_partitions_def.start >= to_partitions_def.start
+                or from_partitions_subset.first_start >= to_partitions_def.start
+            )
+            and (
+                from_last_partition_window is not None
+                and to_last_partition_window is not None
+                and from_last_partition_window.end <= to_last_partition_window.end
+            )
+        ):
+            return UpstreamPartitionsResult(
+                TimeWindowPartitionsSubset(
+                    partitions_def=to_partitions_def,
+                    num_partitions=None,
+                    included_time_windows=from_partitions_subset.included_time_windows,
+                ),
+                [],
+            )
+
+        # The subset we're mapping from doesn't exist in the PartitionsDefinition we're mapping to
+        if from_partitions_subset.cheap_ends_before(
+            to_partitions_def.start, to_partitions_def.cron_schedule
+        ):
+            if self.allow_nonexistent_upstream_partitions:
+                required_but_nonexistent_partition_keys = []
+            else:
+                required_but_nonexistent_partition_keys = [
+                    pk
+                    for time_window in from_partitions_subset.included_time_windows
+                    for pk in to_partitions_def.get_partition_keys_in_time_window(
+                        time_window=time_window
+                    )
+                ]
+
+            return UpstreamPartitionsResult(
+                to_partitions_def.empty_subset(), required_but_nonexistent_partition_keys
+            )
+
+        return None
+
 
 def _offsetted_datetime(
     partitions_def: TimeWindowPartitionsDefinition, dt: datetime, offset: int
-) -> Optional[datetime]:
-    for _ in range(abs(offset)):
-        if offset < 0:
-            prev_window = partitions_def.get_prev_partition_window(dt)
-            if prev_window is None:
-                return None
-
-            dt = prev_window.start
-        else:
-            # TODO: what if we're at the end of the line?
-            next_window = partitions_def.get_next_partition_window(dt)
-            if next_window is None:
-                return None
-
-            dt = next_window.end
-
-    return dt
+) -> datetime:
+    if partitions_def.cron_schedule == "0 0 * * *" and offset != 0:
```

Comment:
> we keep doing these string checks on the cron schedule, it might be worth factoring out into partitions_def.is_default_daily and partitions_def.is_default_hourly or something along those lines.

good idea

Reply:
> also, why not have a similar thing for hourly schedules?

 no good reason. adding.